gr-digital/include/gnuradio/digital/adaptive_algorithm_nlms.h

```diff
@@ -0,0 +1,78 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifndef INCLUDED_DIGITAL_ADAPTIVE_ALGORITHM_NLMS_H
+#define INCLUDED_DIGITAL_ADAPTIVE_ALGORITHM_NLMS_H
+
+#include <gnuradio/digital/adaptive_algorithm.h>
+#include <volk/volk.h>
+
+namespace gr {
+namespace digital {
+
+class DIGITAL_API adaptive_algorithm_nlms : public adaptive_algorithm
+{
+public:
+    typedef boost::shared_ptr<adaptive_algorithm_nlms> sptr;
+
+    float d_step_size;
+
+protected:
+    adaptive_algorithm_nlms(constellation_sptr cons, float step_size)
+        : adaptive_algorithm(adaptive_algorithm_t::NLMS, cons)
+    {
+        d_step_size = step_size;
+    }
+
+public:
+    static sptr make(constellation_sptr cons, float step_size)
+    {
+        return adaptive_algorithm_nlms::sptr(
+            new adaptive_algorithm_nlms(cons, step_size));
+    }
+    void update_tap(gr_complex& tap,
+                    const gr_complex& in,
+                    const gr_complex error,
+                    const gr_complex decision)
+    {
+        throw std::runtime_error(
+            "NLMS can only update all taps at once, single tap update is not valid");
+    }
+
+    void update_taps(gr_complex* taps,
+                     const gr_complex* in,
+                     const gr_complex error,
+                     const gr_complex decision,
+                     unsigned int num_taps)
+    {
+
+        gr_complex dp;
+        volk_32fc_x2_conjugate_dot_prod_32fc(&dp, in, in, num_taps);
+        float magsq = abs(dp);
```

Comment:
You are correct abs is not necessary here - since the dot product does the squaring and x*x' is a positive number, I think we can actually get away with real() here.

Reply:
Yes, you’re right. I did saw that the required dot product with a real result is performed. But then my brain failed and only saw the temptingly wrong looking ``std::abs()``…